File: packages/cli-node/src/commands/reinit.ts
// spec-init reinit [dir]: adopt Throughspec in place, keeping existing files unless --force.

import { existsSync, mkdirSync, readFileSync, writeFileSync } from 'node:fs';
import { dirname, join, relative, resolve } from 'node:path';
import { UsageError } from '../args.js';
import type { CliOptions } from '../args.js';
import { resolvePayloadDir } from '../payload.js';
import { applyConfig, DEFAULT_PERSONA, isManaged } from '../config.js';
import { installIntegrations } from '../integrations.js';
import type { InstallResult } from '../integrations.js';
import { postInitChecklist } from '../checklist.js';
import {
  INTEGRATIONS_PREFIX,
  integrationFilesFor,
  maybeTransform,
  promptIntegrations,
  promptPersona,
  readLineSync,
  walkPayload,
} from './init.js';
import type { PromptDeps } from './init.js';

type ReinitMode = 'keep' | 'replace';

interface ReinitResult {
  dir: string;
  written: number;
  kept: number;
  dryRun: boolean;
}

/** Adopt Throughspec into an existing project in place; quiet skips prompts + output. */
export function runReinit(opts: CliOptions, quiet = false): ReinitResult {
  const [dirArg] = opts.positional;
  const dir = resolve(process.cwd(), dirArg ?? '.');
  const payloadDir = resolvePayloadDir();

  if (isManaged(dir)) {
    throw new UsageError(
      'this project already has spec.config.js - it looks Throughspec-managed.\n' +
        '       Run `spec-init upgrade` to merge a newer template, or `spec-init customize` to change options.',
    );
  }

  const persona = quiet ? opts.persona ?? DEFAULT_PERSONA : promptPersona(opts);
  const integrations = quiet ? [...opts.integrations] : promptIntegrations(opts);

  const baseFiles = walkPayload(payloadDir).filter((rel) => !rel.startsWith(INTEGRATIONS_PREFIX));
  const integrationRel = integrationFilesFor(payloadDir, integrations);
  const existing = new Set(
    [...baseFiles, ...integrationRel].filter((rel) => existsSync(join(dir, rel))),
  );
  const mode = quiet
    ? resolveReinitMode(opts, existing.size, { isTty: false, ask: () => '' })
    : resolveReinitMode(opts, existing.size);

  if (opts.dryRun) {
    printPlan(relative(process.cwd(), dir) || '.', baseFiles, integrationRel, existing, mode);
    return { dir, written: 0, kept: 0, dryRun: true };
  }

  let written = 0;
  let kept = 0;

  for (const rel of baseFiles) {
    const dest = join(dir, rel);
    if (existing.has(rel) && mode === 'keep') {
      kept += 1;
      continue;
    }
    mkdirSync(dirname(dest), { recursive: true });
    let content = maybeTransform(rel, readFileSync(join(payloadDir, rel), 'utf8'), persona, integrations);
    if (rel === 'spec.config.js') content = applyConfig(content, { persona, integrations });
    writeFileSync(dest, content);
    written += 1;
  }

  // Integration files copied raw, honoring the same keep/replace policy.
  for (const name of integrations) {
    const root = join(payloadDir, INTEGRATIONS_PREFIX, name);
    if (!existsSync(root)) continue;
    for (const rel of walkPayload(root)) {
      const dest = join(dir, rel);
      if (existing.has(rel) && mode === 'keep') {
        kept += 1;
        continue;
      }
      mkdirSync(dirname(dest), { recursive: true });
      writeFileSync(dest, readFileSync(join(root, rel)));
      written += 1;
    }
  }

  const installResults: InstallResult[] = quiet
    ? []
    : installIntegrations(integrations, {
        isTty: Boolean(process.stdin.isTTY),
        noInstall: opts.noInstall,
      });

  const relDir = relative(process.cwd(), dir) || '.';
  if (!quiet) {
    process.stdout.write(`\nAdopted Throughspec in ${relDir}: ${written} written, ${kept} kept.\n`);
    process.stdout.write(postInitChecklist(relDir, persona, integrations, installResults, 'Initialized'));
  }
  return { dir, written, kept, dryRun: false };
}

/** Decide keep vs replace for existing spec files. Default is the safe `keep`. */
export function resolveReinitMode(
  opts: CliOptions,
  existingCount: number,
  deps: PromptDeps = { isTty: Boolean(process.stdin.isTTY), ask: readLineSync },
): ReinitMode {
  if (opts.force) return 'replace';
  if (existingCount === 0) return 'replace'; // nothing to keep; write every missing file
  if (opts.dryRun || !deps.isTty) return 'keep'; // safe non-interactive default
  process.stdout.write(
    `Found ${existingCount} existing spec file(s). Keep them, or replace with fresh templates? [keep/replace] `,
  );
  const answer = deps.ask().trim().toLowerCase();
  return answer === 'replace' || answer === 'r' ? 'replace' : 'keep';
}

/** Print the dry-run plan: which files would be written vs kept. */
function printPlan(
  relDir: string,
  baseFiles: readonly string[],
  integrationRel: readonly string[],
  existing: ReadonlySet<string>,
  mode: ReinitMode,
): void {
  const all = [...baseFiles, ...integrationRel];
  const willKeep = mode === 'keep' ? all.filter((rel) => existing.has(rel)) : [];
  const willWrite = all.filter((rel) => !willKeep.includes(rel));
  process.stdout.write(
    `[dry-run] reinit ${relDir} (mode: ${mode}): ${willWrite.length} would be written, ${willKeep.length} kept\n`,
  );
  for (const rel of willWrite) process.stdout.write(`[dry-run]   write  ${rel}\n`);
  for (const rel of willKeep) process.stdout.write(`[dry-run]   keep   ${rel}\n`);
}

File: packages/cli-python/src/spec_init/commands/reinit.py
"""spec-init reinit [dir]: adopt Throughspec in place, keeping existing files unless --force."""

from __future__ import annotations

import json
import shutil
import sys
from collections.abc import Callable
from dataclasses import dataclass
from pathlib import Path
from typing import Literal

from ..args import CliOptions, UsageError
from ..checklist import post_init_checklist
from ..payload import resolve_payload_dir
from .init import (
    INTEGRATIONS_PREFIX,
    _maybe_transform,
    _walk_payload,
    integration_files_for,
    prompt_integrations,
)

ReinitMode = Literal["keep", "replace"]


@dataclass(frozen=True)
class ReinitResult:
    """Summary of a completed reinit run."""

    directory: Path
    written: int
    kept: int
    dry_run: bool


def run_reinit(opts: CliOptions) -> ReinitResult:
    """Adopt Throughspec into an existing project in place. Raises UsageError."""
    dir_arg = opts.positional[0] if opts.positional else "."
    directory = (Path.cwd() / dir_arg).resolve()
    payload_dir = resolve_payload_dir()

    if (directory / ".spec-init" / "base").exists():
        raise UsageError(
            "this project already has a .spec-init/base snapshot - it looks "
            "Throughspec-managed.\n"
            "       Run `spec-init upgrade` to merge a newer template, or "
            "`spec-init customize` to change options."
        )

    integrations = prompt_integrations(opts)

    base_files = [
        rel for rel in _walk_payload(payload_dir) if not rel.startswith(INTEGRATIONS_PREFIX)
    ]
    integration_rel = integration_files_for(payload_dir, integrations)
    existing = {
        rel for rel in (*base_files, *integration_rel) if (directory / rel).exists()
    }
    mode = resolve_reinit_mode(opts, len(existing))

    if opts.dry_run:
        _print_plan(directory, base_files, integration_rel, existing, mode)
        return ReinitResult(directory=directory, written=0, kept=0, dry_run=True)

    written = 0
    kept = 0

    for rel in base_files:
        dest = directory / rel
        if rel in existing and mode == "keep":
            kept += 1
            continue
        dest.parent.mkdir(parents=True, exist_ok=True)
        content = _maybe_transform(
            rel, (payload_dir / rel).read_text(encoding="utf-8"), opts.persona, integrations
        )
        dest.write_text(content, encoding="utf-8", newline="")
        written += 1

    # Integration payload files are copied raw, honoring the same keep/replace policy.
    for name in integrations:
        root = payload_dir / "_integrations" / name
        if not root.exists():
            continue
        for rel in _walk_payload(root):
            dest = directory / rel
            if rel in existing and mode == "keep":
                kept += 1
                continue
            dest.parent.mkdir(parents=True, exist_ok=True)
            dest.write_bytes((root / rel).read_bytes())
            written += 1

    base_dir = directory / ".spec-init" / "base"
    shutil.copytree(payload_dir, base_dir, dirs_exist_ok=True)
    (directory / ".spec-init" / "meta.json").write_text(
        json.dumps({"persona": opts.persona, "integrations": list(integrations)}, indent=2)
        + "\n",
        encoding="utf-8",
    )

    try:
        rel_dir = str(directory.relative_to(Path.cwd())) or "."
    except ValueError:
        rel_dir = str(directory)
    sys.stdout.write(f"\nAdopted Throughspec in {rel_dir}: {written} written, {kept} kept.\n")
    sys.stdout.write(post_init_checklist(rel_dir, opts.persona, integrations, "Initialized"))
    return ReinitResult(directory=directory, written=written, kept=kept, dry_run=False)


def resolve_reinit_mode(
    opts: CliOptions,
    existing_count: int,
    isatty: Callable[[], bool] | None = None,
    ask: Callable[[str], str] | None = None,
) -> ReinitMode:
    """Decide keep vs replace for existing spec files. Default is the safe `keep`."""
    if isatty is None:
        isatty = sys.stdin.isatty
    if ask is None:
        ask = input
    if opts.force:
        return "replace"
    if existing_count == 0:
        return "replace"  # nothing to keep; write every missing file
    if opts.dry_run or not isatty():
        return "keep"  # safe non-interactive default
    answer = ask(
        f"Found {existing_count} existing spec file(s). Keep them, or replace "
        "with fresh templates? [keep/replace] "
    ).strip().lower()
    return "replace" if answer in ("replace", "r") else "keep"


def _print_plan(
    directory: Path,
    base_files: list[str],
    integration_rel: list[str],
    existing: set[str],
    mode: ReinitMode,
) -> None:
    """Print the dry-run plan: which files would be written vs kept."""
    all_files = [*base_files, *integration_rel]
    will_keep = [rel for rel in all_files if rel in existing] if mode == "keep" else []
    keep_set = set(will_keep)
    will_write = [rel for rel in all_files if rel not in keep_set]
    try:
        rel_dir = str(directory.relative_to(Path.cwd())) or "."
    except ValueError:
        rel_dir = str(directory)
    sys.stdout.write(
        f"[dry-run] reinit {rel_dir} (mode: {mode}): "
        f"{len(will_write)} would be written, {len(will_keep)} kept\n"
    )
    for rel in will_write:
        sys.stdout.write(f"[dry-run]   write  {rel}\n")
    for rel in will_keep:
        sys.stdout.write(f"[dry-run]   keep   {rel}\n")
